### DISCUSSION TITLE
is there any way to copy the contents (like sh files) into final image ?

### DISCUSSION THREAD

**Discussion by Rvarman0110** [paketo-buildpacks/feedback#254] [Java Team]
Hi, I am having a java code using buildpacks I was able to build it but in that source code itself we have some script files .sh files I want it to be there in the final image but I couldn't able to do that is there any way that I can include the files that is needed to be present in the final image.

Like the COPY command in dockerfile is there any similar stuff for Buildpacks ?

**Comment by dmikusa** [comment]
By default, when you do a source build the buildpack will remove all of the non-compiled bits (i.e. everything not in your WAR/JAR file). This makes for the smallest image possible. 

If you want to keep stuff, you have two options:

1. Put the files you want to keep into the WAR/JAR file.
2. You can use the `BP_INCLUDE_FILES` or `BP_EXCLUDE_FILES` env variables to tell it what to include/exclude. See for more details: https://github.com/paketo-buildpacks/maven?tab=readme-ov-file#configuration

Just keep in mind that when using the include/exclude option the path does not change on any of the resources that are include/excluded. So if you have something like `/workspace/src/main/resources/foo.sh` and you tell the buildpack to keep it, the path is the same. It is a common misconception that it would get move to `/workspace/foo.sh`. That won't happen.

If you want the path to change, then you need to put the resources into your WAR/JAR file. After build, files are removed and the WAR/JAR is exploded out into `/workspace` so you if you need to control where a script ends up that's how you'd go about it.

Hope that helps!

**Reply by jrrajav** [reply]
@dmikusa 
Hi

Say I have a target/classes folder 

and in the final image , I only want the classes folder
But if I uses BP_INCLUDE_FILES=target/classes

it is including the entire directory 

anyway to include only the classes folder into final image without the target(parent directory)

Like in docker file 
COPY target/classes classes

**Reply by dmikusa** [reply]
@jrrajav We don't presently support that with the `BP_INCLUDE_FILES`/`BP_EXCLUDE_FILES`. I'm not opposed to adding it, but I'm not totally certain how we'd do that in a way that we can expose through env variable configuration and not have it be super complicated. If you have ideas, feel free to post them.

What I would suggest instead is to add your own buildpack. You can use an [inline buildpack](https://buildpacks.io/docs/for-app-developers/how-to/build-inputs/use-inline-buildpacks/) which is a light, build-only, single script, or you can [bundle up your own buildpack](https://buildpacks.io/docs/for-buildpack-authors/tutorials/basic-buildpack/), which isn't too much more work (two scripts instead of one). In either case, you'll get a chance to run some custom code so you can then move/copy files around.

The only other thing I want to mention is that the default behavior of Java buildpacks is to take what's bundled in your JAR/WAR, save that, delete `/workspace/*`, then restore the contents of your JAR/WAR to `/workspace`. tl;dr your JAR/WAR is usually going to have `target/classes` in it, so you might already have those files in `/workspace` without the need to include/exclude anything, or make custom buildpack scripts.

If for some reason the file you want are not in your JAR/WAR, another option is to put them in your JAR/WAR at the path where you'd like them to appear in `/workspace`. If you for example put something into the JAR/WAR  at `foo/bar`, then you'll have `/workspace/foo/bar` after the buildpacks run.

hope that helps!

**Comment by llint** [comment]
It seems that `BP_INCLUDE_FILES` does not have any effect within a Maven pom.xml file using the spring-boot-maven-plugin to build an OCI image?

```
					<plugin>
						<groupId>org.springframework.boot</groupId>
						<artifactId>spring-boot-maven-plugin</artifactId>
						<executions>
							<execution>
								<goals>
									<goal>build-image</goal>
								</goals>
							</execution>
						</executions>
						<configuration>
							<image>
								<env>
									<BP_INCLUDE_FILES>my-custom-resources/*</BP_INCLUDE_FILES>
								</env>
							</image>
						</configuration>
					</plugin>

```

**Reply by dmikusa** [reply]
It should work. There's nothing that is specific to the Spring Boot Build Tools plugin and the way it runs builds with Paketo buildpacks that would impact this functionality. If you're having issues, I suggest you open a new thread and expand on the problem you're seeing.

**Reply by ozozgun** [reply]
I confirm trying to copy a file from project root does not work.
However if you try to copy a file from resources directory, the file will be copied under `workspace/BOOT-INF/classes`
Which is documented nowhere.
Example:
```
<plugin>
    <groupId>org.springframework.boot</groupId>
    <artifactId>spring-boot-maven-plugin</artifactId>
    <configuration>
        <image>
            <builder>paketobuildpacks/builder-jammy-base:latest</builder>
            <env>
                <BP_INCLUDE_FILES>src/main/resources/org.jacoco.agent.jar</BP_INCLUDE_FILES>
            </env>
         </image>
    </configuration>
</plugin>
```
And you will find your file on your container:
```
cnb@36865f994a2d:/workspace$ ls BOOT-INF/classes/
application-batch.yaml  application-gcp.yaml  application-maintenance.yaml  application.yaml  com  org.jacoco.agent.jar  spy.properties
cnb@36865f994a2d:/workspace$ 
```

FYI, I use spring-boot-maven-plugin-3.5.8
